pkg/cloudprovider/provider/vmware-cloud-director/client.go

```diff
@@ -0,0 +1,173 @@
+	"net/url"
+	"strings"
+
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+
+	cloudprovidererrors "github.com/kubermatic/machine-controller/pkg/cloudprovider/errors"
+)
+
+type Client struct {
+	Auth      *Auth
+	VCDClient *govcd.VCDClient
+}
+
+func NewClient(username, password, org, url, vdc string, allowInsecure bool) (*Client, error) {
+	client := Client{
+		Auth: &Auth{
+			Username:      username,
+			Password:      password,
+			Organization:  org,
+			URL:           url,
+			VDC:           vdc,
+			AllowInsecure: allowInsecure,
+		},
+	}
+
+	vcdClient, err := client.GetAuthenticatedClient()
+	if err != nil {
+		return nil, err
+	}
+
+	client.VCDClient = vcdClient
+	return &client, nil
+}
+
+func (c *Client) GetAuthenticatedClient() (*govcd.VCDClient, error) {
+	// Ensure that all required fields for authentication are provided
+	// Fail early, without any API calls, if some required field is missing.
+	if c.Auth == nil {
+		return nil, fmt.Errorf("authentication configuration not provided")
+	}
+	if c.Auth.Username == "" {
+		return nil, fmt.Errorf("username not provided")
+	}
+	if c.Auth.Password == "" {
+		return nil, fmt.Errorf("password not provided")
+	}
+	if c.Auth.URL == "" {
+		return nil, fmt.Errorf("URL not provided")
+	}
+	if c.Auth.Organization == "" {
+		return nil, fmt.Errorf("organization name not provided")
+	}
+
+	// Ensure that `/api` suffix exists in the cloud director URL.
+	href := c.Auth.URL
+	if !strings.HasSuffix(c.Auth.URL, "/api") {
+		href = fmt.Sprintf("%s/api", c.Auth.URL)
+	}
+
+	url, err := url.ParseRequestURI(href)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse url '%s': %w", href, err)
+	}
+
+	vcdClient := govcd.NewVCDClient(*url, c.Auth.AllowInsecure)
+
+	err = vcdClient.Authenticate(c.Auth.Username, c.Auth.Password, c.Auth.Organization)
+	if err != nil {
+		return nil, fmt.Errorf("failed to authenticate with VMware Cloud Director: %w", err)
+	}
+
+	return vcdClient, nil
+}
+
+func (c *Client) GetOrganization() (*govcd.Org, error) {
+	if c.Auth.Organization == "" {
+		return nil, errors.New("organization must be configured")
+	}
+
+	org, err := c.VCDClient.GetOrgByNameOrId(c.Auth.Organization)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get organization '%s': %w", c.Auth.Organization, err)
+	}
+	return org, err
+}
+
+func (c *Client) GetVDCForOrg(org govcd.Org) (*govcd.Vdc, error) {
+	if c.Auth.VDC == "" {
+		return nil, errors.New("Organization VDC must be configured")
+	}
+	vcd, err := org.GetVDCByNameOrId(c.Auth.VDC, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Organization VDC '%s': %w", c.Auth.VDC, err)
+	}
+	return vcd, err
+}
+
+func (c *Client) GetVMByName(vappName, vmName string) (*govcd.VM, error) {
+	_, _, vapp, err := c.GetOrganizationVDCAndVapp(vappName)
+	if err != nil {
+		return nil, err
+	}
+
+	// We don't need ID here since we explicitly set the name field when creating the resource.
```

Comment:
Yes, but not sure that there is much that we can do about it. In this context, since the VM is created using `machine-controller` we assume that we can just use the name.

Reply:
What if we have more than one Machine with the same name? Which one will be picked by the client? Why don't we use the ID instead?

Reply:
We can't set the `ID` when creating a VM. It's auto-assigned.

Reply:
We decided to leave this for a follow-up.